docs/topics/proc_exposing-application-metrics-using-prometheus-on-node-js.adoc

```diff
@@ -0,0 +1,134 @@
++
+[source,yaml]
+----
+  - job_name: nodejs
+     static_configs:
+     - targets: ['localhost:8888']
+----
++
+The default behavior of {Node}-based applications is to expose metrics at the `/metrics` endpoint.
+This is also what Prometheus expects.
+--
+
+* The Prometheus server running on `localhost` using Docker:
++
+--
+[source,bash,opts="nowrap"]
+----
+# docker run -it --rm --network host -p 9090:9090 -v $PWD/prometheus.yml:/etc/prometheus/prometheus.yml prom/prometheus
+----
+
+Wait until the `Server is ready to receive web requests` message is displayed in the console.
+
+When running your application in a Linux environment, include the the `--network host` option to allow Prometheus to connect to your application.
+
+NOTE: The `--network host` option is available only in Linux environments.
+It is not supported on Windows or MacOS. 
+--
+
+.Procedure
+
+. Install the `prom-client` npm package.
++
+.pom.xml
+[source,bash]
+----
+$ npm install prom-client express body-parser --save
+----
+
+. Create an example {NodeJS} application and save it in a file with the file name `app.js` .
++
+--
+[source,javascript]
+----
+'use strict';
+const path = require('path');
+const express = require('express');
+const bodyParser = require('body-parser');
+const client = require('prom-client');
+
+// enable prom-client to expose default application metrics
+const collectDefaultMetrics = client.collectDefaultMetrics;
+
+// define a custom prefix string for application metrics
+collectDefaultMetrics({ prefix: 'application:' });
+
+const histogram = new client.Histogram({
+  name: 'application:hello',
+  help: 'Duration of HTTP requests in ms',
+  labelNames: ['status_code'],
+  buckets: [0.1, 5, 15, 50, 100, 500]
+});
+const app = express();
+const port = 8888;
+app.use(bodyParser.json());
+app.use(bodyParser.urlencoded({extended: false}));
+app.use('/api/greeting', (request, response) => {
+  const end = histogram.startTimer();
+  const name = request.query ? request.query.name : undefined;
+  response.send({content: `Hello, ${name || 'World!'}`});
+  end();
+});
+
+app.get('/metrics', (req, res) => {
+  res.set('Content-Type', client.register.contentType);
+  res.end(client.register.metrics());
+})
+
+app.listen(port, () => console.log(`Example app listening on port ${port}!`))
+----
+--
+
+. Launch your application:
++
+[source,bash,opts="nowrap"]
+----
+$ node app.js
+----
+
+. Invoke the traced endpoint several times:
++
+[source,bash,opts="nowrap"]
+----
+$ curl http://localhost:8888/api/greeting
+{"content":"Hello, World!"}
+----
+
+. Wait at least 15 seconds for the collection to happen, and view the metrics in the Prometheus web UI:
++
+--
+. Open the Prometheus UI at link:http://localhost:9090/[http://localhost:9090/^] and type `hello` into the _Expression_ box.
+. From the suggestions, select for example `application:hello_count` and click the _Execute_ button.
+Note that all metrics you created are prefixed with `application:`.
```

Comment:
This is what I've been using locally on mac:
```yml
- job_name: 'nodejs'
    static_configs:
    - targets: ['192.168.1.80:8888']
```
```console
$ docker run -it --rm -p 9090:9090 -v $PWD/prom.yml:/etc/prometheus/prometheus.yml prom/prometheus
```
And then updating the app to listen on the ip address:
```javascript
'use strict';

const path = require('path');
const express = require('express');
const bodyParser = require('body-parser');
const client = require('prom-client');

const collectDefaultMetrics = client.collectDefaultMetrics;
collectDefaultMetrics({ prefix: 'application:' });

const histogram = new client.Histogram({
  name: 'application:hello',
  help: 'Duration of HTTP requests in ms',
  labelNames: ['status_code'],
  // buckets for response time from 0.1ms to 500ms
  buckets: [0.1, 5, 15, 50, 100, 500]
});

const app = express();
const port = 8888;

app.use(bodyParser.json());
app.use(bodyParser.urlencoded({extended: false}));

app.use('/api/greeting', (request, response) => {
  const end = histogram.startTimer();
  const name = request.query ? request.query.name : undefined;
  response.send({content: `Hello, ${name || 'World!'}`});
  end();
});

app.get('/metrics', (req, res) => {
  res.set('Content-Type', client.register.contentType);
  res.end(client.register.metrics());
})
const ip = '192.168.1.80';
app.listen(port, ip, () => console.log(`Example app listening on port ${ip}:${port}!`))
```